lib/Model/UserConfig.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OCA\Gitlab\Model;
+
+use InvalidArgumentException;
+use OCA\Gitlab\Service\ConfigService;
+use OCP\PreConditionNotMetException;
+
+class UserConfig {
+	private function __construct(
+		public ?string $token,
+		public ?string $url,
+		public ?string $client_id,
+		// don't expose the client secret to users
+		public ?bool $client_secret,
+		public ?string $oauth_instance_url,
+		public ?bool $use_popup,
+		public ?string $user_name,
+		public ?string $user_displayname,
+		public ?bool $search_enabled,
+		public ?bool $search_issues_enabled,
+		public ?bool $search_mrs_enabled,
+		public ?bool $navigation_enabled,
+		public ?bool $link_preview_enabled,
+		public ?string $oauth_state,
+		public ?string $redirect_uri,
+		public ?string $oauth_origin,
+	) {
+	}
+
+	public static function fromConfig(string $userId, ConfigService $config): UserConfig {
+		return new UserConfig(
+			token: $config->getUserToken($userId),
+			url: $config->getUserUrl($userId),
+			client_id: $config->getAdminClientId(),
+			client_secret: $config->hasAdminClientSecret(),
+			oauth_instance_url: $config->getAdminOauthUrl(),
+			use_popup: $config->getAdminUsePopup(),
+			user_name: $config->getUserName($userId),
+			user_displayname: $config->getUserDisplayName($userId),
+			search_enabled: $config->getUserSearchEnabled($userId),
+			search_issues_enabled: $config->getUserSearchIssuesEnabled($userId),
+			search_mrs_enabled: $config->getUserSearchMergeRequestsEnabled($userId),
+			navigation_enabled: $config->getUserNavigationEnabled($userId),
+			link_preview_enabled: $config->getUserLinkPreviewEnabled($userId),
+			oauth_state: $config->getUserOauthState($userId),
+			redirect_uri: $config->getUserRedirectUri($userId),
+			oauth_origin: $config->getUserOauthOrigin($userId),
+		);
+	}
+
+	/**
+	 * @throws PreConditionNotMetException
+	 */
+	public function toConfig(string $userId, ConfigService $config): void {
+		if ($this->client_id !== null || $this->client_secret !== null || $this->oauth_instance_url !== null || $this->use_popup !== null || $this->user_name !== null || $this->user_displayname !== null) {
+			throw new InvalidArgumentException();
```

Comment:
also throws `InvalidArgumentException`

Reply:
TBH I did this on purpose because the exception is only thrown when there is a programming error and thus it should never occur and be catched. If you want I can change it though

Reply:
yeah I would suggest to add it just for the users of function if they for any purpose want to know the exceptions thrown. A rarity but doesn't hurt.